src/ewah.zig

```diff
@@ -61,98 +64,208 @@
+
+            /// Returns the number of *words* written to `target_words` by this invocation.
+            // TODO Refactor to return an error when `source_chunk` is invalid,
+            // so that we can test invalid encodings.
+            pub fn decode(
+                decoder_: *Decoder,
+                source_chunk: []align(@alignOf(Word)) const u8,
+            ) usize {
+                assert(source_chunk.len % @sizeOf(Word) == 0);
+
+                const source_words = mem.bytesAsSlice(Word, source_chunk);
+                const target_words = decoder_.target_words;
+                assert(disjoint_slices(u8, Word, source_chunk, target_words));
+
+                var source_index: usize = 0;
+                var target_index: usize = decoder_.target_index;
+
+                if (decoder_.source_literal_words > 0) {
+                    const literal_word_count_chunk =
+                        @min(decoder_.source_literal_words, source_words.len);
+
+                    stdx.copy_disjoint(
+                        .exact,
+                        Word,
+                        target_words[target_index..][0..literal_word_count_chunk],
+                        source_words[source_index..][0..literal_word_count_chunk],
+                    );
+                    source_index += literal_word_count_chunk;
+                    target_index += literal_word_count_chunk;
+                    decoder_.source_literal_words -= literal_word_count_chunk;
+                }
+
+                while (source_index < source_words.len and target_index < target_words.len) {
+                    assert(decoder_.source_literal_words == 0);
+
+                    const marker: *const Marker = @ptrCast(&source_words[source_index]);
+                    source_index += 1;
+                    @memset(
+                        target_words[target_index..][0..marker.uniform_word_count],
+                        if (marker.uniform_bit == 1) ~@as(Word, 0) else 0,
+                    );
+                    target_index += marker.uniform_word_count;
+
+                    const literal_word_count_chunk =
+                        @min(marker.literal_word_count, source_words.len - source_index);
+                    stdx.copy_disjoint(
+                        .exact,
+                        Word,
+                        target_words[target_index..][0..literal_word_count_chunk],
+                        source_words[source_index..][0..literal_word_count_chunk],
+                    );
+                    source_index += literal_word_count_chunk;
+                    target_index += literal_word_count_chunk;
+                    decoder_.source_literal_words =
+                        marker.literal_word_count - literal_word_count_chunk;
+                }
+                assert(source_index <= source_words.len);
+                assert(target_index <= target_words.len);
+
+                defer decoder_.target_index = target_index;
```

Comment:
Nifty! But maybe move this defer right to the `var target_index` declaration, so that the reader
immediately knows the final destination of that local variable?